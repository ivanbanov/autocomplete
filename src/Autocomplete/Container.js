// @flow

import styled from 'styled-components';
import { type ComponentType } from 'react';

type Props = {};

const Container: ComponentType<Props> =
  styled.div`
    position: relative;
  `;

Container.displayName = 'Autocomplete.Container';

export default Container;
