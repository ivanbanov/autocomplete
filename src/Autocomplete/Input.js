// @flow

import styled from 'styled-components';
import { type ComponentType } from 'react';

type Props = {
  fluid?: boolean,
};

const Input: ComponentType<Props> =
  styled.input.attrs({
    type: 'search',
  })`
    border: 1px solid #ddd;
    padding: 1rem;
    width: ${props => (props.fluid ? '100%' : 'auto')};
    outline: none;
    border-radius: 3px;

    &:focus {
      border-color: #bbb;
    }
  `;

Input.displayName = 'Autocomplete.Input';

export default Input;
