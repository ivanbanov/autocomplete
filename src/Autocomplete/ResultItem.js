// @flow

import styled from 'styled-components';
import { type ComponentType } from 'react';

type Props = {
  isActive: boolean,
};

const ResultItem: ComponentType<Props> =
  styled.div`
    padding: .5rem;
    background: ${props => (props.isActive ? '#ddd' : '#fff')};

    & + & { padding-top: 1px solid #ddd; }

    &:hover {
      background: #ddd;
    }
  `;

ResultItem.displayName = 'Autocomplete.ResultItem';

export default ResultItem;
