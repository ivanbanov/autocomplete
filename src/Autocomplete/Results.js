// @flow

import styled from 'styled-components';
import { type ComponentType } from 'react';

type Props = {};

const Results: ComponentType<Props> =
  styled.div`
    position: absolute;
    top: 100%;
  `;

Results.displayName = 'Autocomplete.Results';

export default Results;
