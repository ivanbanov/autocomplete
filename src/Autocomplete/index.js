// @flow

import React, { type Element } from 'react';
import removeAccents from 'remove-accents';
import Container from './Container';
import Input from './Input';
import Results from './Results';
import ResultItem from './ResultItem';

const NO_RESULTS = 'No results';

type Props = {
  data: Array<string> | string,
  value?: string,
  parseResults?: Function,
  maxItems?: number,
};

type State = {
  results: Array<*>,
  value: ?string,
  isLoading: boolean,
  activeIndex: number,
}

class Autocomplete extends React.Component<Props, State> {
    static displayName = 'Autocomplete';

    static props: Props;

    state: State = {
      results: [],
      value: null,
      isLoading: false,
      activeIndex: -1,
    };

    constructor(props: Props) {
      super(props);

      this.state.value = this.props.value || '';

      (this: any).onBlur = this.onBlur.bind(this);
      (this: any).onChange = this.onChange.bind(this);
      (this: any).onKeyDown = this.onKeyDown.bind(this);
      (this: any).renderItem = this.renderItem.bind(this);
      (this: any).getResults = this.getResults.bind(this);
      (this: any).isRemoteData = this.isRemoteData.bind(this);
    }

    getResults: Function;
    input: ?HTMLInputElement;

    isRemoteData() {
      const { data } = this.props;
      return typeof data === 'string' && data.match(/^https?:\/\/|\/\//);
    }

    getResults(): (Promise<*> | Array<*> | null) {
      const search = (this.input && this.input.value) || '';
      const { data = [] } = this.props;

      // flowtype forces to check typeof again
      if (this.isRemoteData && typeof data === 'string') {
        return fetch(data + search).then(res => res.json());
      } else if (Array.isArray(data)) {
        return data;
      }

      return null;
    }

    onChange(e: any): void {
      const {
        data = [],
        parseResults,
        maxItems = 5,
      } = this.props;

      const { value } = e.target;

      this.setState({ value });

      if (this.isRemoteData()) {
        this.setState({
          isLoading: true,
          results: ['Loading...'],
        });

        this.getResults().then(json => {
          let results = parseResults ? parseResults(json) : [];

          if (!results.length) {
            results = [NO_RESULTS];
          }

          this.setState({
            isLoading: false,
            results: results.slice(0, maxItems),
          });
        });

        return;
      }

      const dataArr = Array.isArray(data) ? data : [];
      let filteredResults = value === ''
        ? []
        : dataArr
          .filter(item => {
            const regexp = new RegExp(removeAccents(value), 'ig');
            return regexp.test(item);
          })
          .slice(0, maxItems);

      if (!filteredResults.length) {
        filteredResults = [NO_RESULTS];
      }

      this.setState({ results: filteredResults });
    }

    onKeyDown(e: any): void {
      const { activeIndex, results } = this.state;
      const { value } = e.target;

      if (!results.length) {
        return;
      }

      switch (e.key) {
        case 'ArrowUp':
          this.setState({
            activeIndex: (activeIndex === 0) ? 0 : (activeIndex - 1),
          });
          break;

        case 'ArrowDown':
          this.setState({
            activeIndex: (activeIndex === results.length - 1) ? activeIndex : (activeIndex + 1),
          });
          break;

        case 'Enter':
          this.setState({
            activeIndex: -1,
            results: [],
            value: results[activeIndex],
          });
          break;

        case 'Escape':
          this.setState({
            activeIndex: -1,
            results: [],
          });
          break;

        default:
          this.setState({ value });
      }
    }

    onBlur(): void {
      setTimeout(() => {
        this.setState({ results: [] });
      }, 50);
    }

    renderItem(item: string, key: number = 1): Element<*> {
      return (
        <ResultItem
          key={key}
          onClick={() => this.setState({ value: item })}
          isActive={key === this.state.activeIndex}
        >
          {item}
        </ResultItem>
      );
    }

    render() {
      const {
        results,
        value,
      } = this.state;

      return (
        <Container>
          <Input
            onChange={this.onChange}
            onBlur={this.onBlur}
            onKeyDown={this.onKeyDown}
            innerRef={input => this.input = input}
            value={value}
          />
          {
            !!results.length &&
            <Results>
              { results.map(this.renderItem) }
            </Results>
          }
        </Container>
      );
    }
}

export default Autocomplete;
