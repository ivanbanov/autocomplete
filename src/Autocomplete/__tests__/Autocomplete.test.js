import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Autocomplete from '../';
import Input from '../Input';
import Results from '../Results';
import ResultItem from '../ResultItem';

Enzyme.configure({ adapter: new Adapter() });

function simulateChange(component, str) {
  component.find(Input).simulate('change', {
    target: { value: str },
  });
}

describe('Autocomplete', () => {
  it('should render properly', () => {
    const component = renderer.create(
      <Autocomplete data={[]} />
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render with initial value', () => {
    const component = shallow(
      <Autocomplete data={['foo']} value="foo" />
    );

    expect(component.find(Input).dive().find('input').prop('value')).toEqual('foo');
  });

  it('should render results', () => {
    const component = shallow(
      <Autocomplete data={['foo', 'bar']} />
    );

    component.find(Input).simulate('change', {
      target: { value: 'foo' },
    });

    expect(component.find(Results)).toHaveLength(1);
    expect(component.find(Results).dive().find(ResultItem)).toHaveLength(1);
  });

  it('should render results', () => {
    const component = shallow(
      <Autocomplete data={['foo', 'bar']} />
    );

    simulateChange(component, 'foo');

    expect(component.find(Results)).toHaveLength(1);
    expect(component.find(Results).dive().find(ResultItem)).toHaveLength(1);
  });

  it('should render only the max items allowed', () => {
    const component = shallow(
      <Autocomplete data={['foo', 'foo']} maxItems={1} />
    );

    simulateChange(component, 'foo');

    expect(component.find(Results).dive().find(ResultItem)).toHaveLength(1);
  });

  it('should run the onSelect callback', () => {
    const fn = jest.fn();

    const component = shallow(
      <Autocomplete data={['foo', 'bar']} onSelect={fn} />
    );

    simulateChange(component, 'foo');
    component.find(ResultItem).first().simulate('click');

    expect(fn).toHaveBeenCalled();
  });

  it('should navigate with keyboard', () => {
    const fn = jest.fn();

    const component = shallow(
      <Autocomplete
        data={['foo1', 'foo2']}
        onSelect={value => {
          expect(value).toEqual('foo');
          fn();
        }}
      />
    );

    simulateChange(component, 'foo');

    component.find(Input).simulate('keyDown', {
      key: 'ArrowDown', target: { value: 'foo' },
    });
    component.find(Input).simulate('keyDown', {
      key: 'ArrowDown', target: { value: 'foo' },
    });

    expect(component.find(ResultItem).at(1).prop('isActive')).toEqual(true);

    component.find(Input).simulate('keyDown', {
      key: 'ArrowUp', target: { value: 'foo' },
    });

    expect(component.find(ResultItem).at(0).prop('isActive')).toEqual(true);

    component.find(Input).simulate('keyDown', {
      key: 'Enter', target: { value: 'foo' },
    });

    expect(fn).toHaveBeenCalled();
  });
});
