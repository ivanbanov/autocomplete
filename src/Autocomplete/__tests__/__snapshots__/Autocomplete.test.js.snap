// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete should render properly 1`] = `
<div
  className="Container-gZqETv gARBHO"
>
  <input
    className="Input-hxTtdt iQCqrY"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    type="search"
    value=""
  />
</div>
`;
